torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -444,7 +444,61 @@ ExprHandle promoteIntegerToFloat(const ExprHandle& e) {
   return Cast::make(Dtype(defaultType, e.dtype().lanes()), e);
 }
 
-void TensorExprKernel::promoteInputs(std::vector<ExprHandle>& inputs) {
+ExprHandle promoteHalfAndIntegerToFloat(const ExprHandle& e) {
+  auto scalarType = static_cast<c10::ScalarType>(e.dtype().scalar_type());
```

Comment:
Not for this diff but this reminds me that I'd love for tensorexpr::Dtype::scalar_type to just use c10::ScalarType instead of defining its own thing that's kind of an extension of it.  Maybe file an issue/task for this?

Reply:
Opened #48784 to track this request. I do not completely understand the scope of this change or the rationale of introducing `tensorexpr::ScalarType` in the first place, feel free to revise the description in #48784 (and add appropriate labels, etc)

Reply:
That's a great description, thanks.  I think the primary reason for te::ScalarType is historical; we started developing out-of-tree, with a group of folks pretty unfamiliar with PyTorch, and later patched in full support for at::ScalarType.  I'm not sure where all we use the "extended" types like kHandle, but I think it'd be cleaner to not pack them into a te-internal ScalarType.  It'd definitely be a somewhat involved change, though limited to tensorexpr/.